075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,424 @@
+     * the `scaleDown()` method skips doing the scale down because a rebalancing is needed first (and the check failed).
+     * the `autoRebalancing()` method detects that a scaling down was requested because of `removedNodes` not being empty but containing the brokers' IDs to be removed from the cluster. Because the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready` (from a previous auto-rebalancing), it's a new operation so a new `KafkaRebalance` custom resource is created, with the `strimzi.io/rebalance-auto-approval: true` annotation, and the `Kafka.status.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`) and the `brokers` with the brokers' IDs.
+2. From the **2nd** to the **(N-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the checks on scaling down the cluster keeps failing but it keeps reverting the scaling operation.
+     * the `scaleDown()` method skips doing scale down because a rebalancing is needed first (and the check failed).
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.autoRebalance.state` field.
+3. On the **Nth** reconciliation (rebalancing id done, scaling down can start):
+     * the checks on scaling down the cluster don't fail, because the auto-rebalancing was completed and the nodes to be removed don't host any partition replicas anymore.
+     * the `scaleDown()` method can run the scaling down.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed). The end of rebalancing is reflected into the `Kafka.status.autoRebalance.state` field and the `KafkaRebalance` is deleted.
+
+During the above flow, errors can happen before the scaling down:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.autoRebalance.state` field.
+
+The reconcile ends but the scale down cannot proceed because the brokers to be removed are still hosting some partition replicas.
+
+On the next reconciliation, the cluster operator will try to do a scale down again, so running the flow as before.
+The previous `KafkaRebalance` custom resource in the `NotReady` state is overridden to generate a new optimization proposal.
```

Comment:
If I am understanding correctly this is different behaviour to the scale up flow. In that flow if something goes wrong it's left completely to the user to manually retry the rebalance, however here the operator retries automatically. I assume the reason is because the scale down hasn't happened yet, so the operator has to keep trying. However, I wonder if having the different behaviour is confusing. Maybe it would be better to fail/block the scale down if the rebalance fails, we could follow what happens today if you try to scale down when there are still partitions on the broker you are removing. Then similar to scale up the user just has to manually sort the rebalance before trying the scale down again

Reply:
That makes more sense to me, it gives the user two options if an error happens, either they delete the KR and let the operator try again, or they handle it themselves. And means it's a similar process for both scale up and down which is less confusing.